lib/os_io.c

```diff
@@ -37,6 +37,130 @@ ___io_module ___io_mod =
 
 /*---------------------------------------------------------------------------*/
 
+/* poll dynamic fdset memory management. */
+
+#ifdef USE_poll
+#define ___fdset_state_size(ps)     ps->os.fdset.size
+#define ___fdset_state_readfds(ps)  ps->os.fdset.readfds
+#define ___fdset_state_writefds(ps) ps->os.fdset.writefds
+
+static int ___fdset_heap_overflow;
+
+static int ___fdset_state_init (___processor_state ps)
+{
+  void *readfds = NULL;
+  void *writefds = NULL;
+
+  readfds  = ___ALLOC_MEM (MAX_CONDVARS/8);
+  if (!readfds)
+    goto error;
+
+  writefds = ___ALLOC_MEM (MAX_CONDVARS/8);
+  if (!writefds)
+    goto error;
+
+  memset (readfds, 0, MAX_CONDVARS/8);
+  memset (writefds, 0, MAX_CONDVARS/8);
+  ___fdset_state_readfds (ps) = readfds;
+  ___fdset_state_writefds (ps) = writefds;
+  ___fdset_state_size (ps) = MAX_CONDVARS;
+
+  return 0;
+
+ error:
```

Comment:
Indentation is shallow here... so OK.  Anyway, general rule is avoid gotos unless it is awkward to do otherwise.